### PR TITLE
ruby_requirement: check for ruby.h

### PR DESCRIPTION
**Pull request by carlorosati**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Linuxbrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Linuxbrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Addresses https://github.com/Linuxbrew/homebrew-core/issues/1798

### CODE REVIEW

File: Library/Homebrew/requirements/ruby_requirement.rb [hunk 1]
```diff
@@ -12,7 +12,20 @@ def initialize(tags)
     which_all("ruby").detect do |ruby|
       version = /\d\.\d/.match Utils.popen_read(ruby, "--version")
       next unless version
-      Version.create(version.to_s) >= Version.create(@version)
+      version = version.to_s
+      if version.to_f > 1.9
```
Comment (sjackman) [inline Library/Homebrew/requirements/ruby_requirement.rb]: Brew requires Ruby > 2.0, so you can skip this check and assume Ruby > 2.0

File: Library/Homebrew/requirements/ruby_requirement.rb [hunk 2]
```diff
@@ -12,7 +12,20 @@ def initialize(tags)
...
+        rubyhdrdir = Utils.popen_read(ruby, "-rrbconfig", "-e", "print RbConfig::CONFIG['rubyhdrdir']")
+        arch = Utils.popen_read(ruby, "-rrbconfig", "-e", "print RbConfig::CONFIG['arch']")
```
Comment (sjackman) [inline Library/Homebrew/requirements/ruby_requirement.rb]: Do you have to check under `arch`? On my system `ruby.h` is in `rubyhdrdir`.

Comment (sjackman) [reply Library/Homebrew/requirements/ruby_requirement.rb]: Consider
```ruby
next unless (rubyhdrdir/"ruby.h").exist?
```

**Review by sjackman — CHANGES_REQUESTED** [review]

File: Library/Homebrew/requirements/ruby_requirement.rb (hunk 1 above)
Comment (carlorosati) [reply Library/Homebrew/requirements/ruby_requirement.rb]: Brew might require Ruby > 2.0, but it is packaged with it's own portable-ruby. The ruby that may be visible in the PATH, can still be 1.8

Comment (sjackman) [reply Library/Homebrew/requirements/ruby_requirement.rb]: We can assume that there is at least Ruby 2.0 installed in addition to other Rubies. I'd prefer not to return that the requirement is satisfied if the Ruby ≥ 2.0 does not have `ruby.h` installed. In that case I'd rather install the brewed Ruby.

Comment (carlorosati) [reply Library/Homebrew/requirements/ruby_requirement.rb]: Assuming that Ruby > 2.0 is installed, how would we access the portable-ruby package with brew? Could I just require rbconfig in ruby_requirements.rb? How would be be able to build against the portable-ruby without it being in the PATH?

Comment (sjackman) [reply Library/Homebrew/requirements/ruby_requirement.rb]: You could add the portable Ruby to the `PATH`, either at the user's command shell prompt, or see for example
https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/utils.rb#L196
`RUBY_PATH` is the path to the Ruby executable.
`RUBY_BIN` is its directory.
See https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/global.rb#L26-L27

**Comment by sjackman** [comment]
Note that `RubyRequirement` has change in Homebrew/brew 1.1.11, which hasn't yet been merged into Linuxbrew. See https://github.com/Homebrew/brew/pull/2083/files

**Comment by carlorosati** [comment]
Do all brew installations come with a portable-ruby? This is all pointless if so.

**Comment by bw-yuhogg** [comment]
> Do all brew installations come with a portable-ruby? This is all pointless if so.

No. Mine does not, for example.

It will be installed automatically if your system Ruby is not new enough, but it is by no means installed on all Linuxbrew installations.

**Comment by sjackman** [comment]
The brewed portable Ruby is used on systems whose host Ruby is < 2.0.

**Comment by carlorosati** [comment]
@sjackman 

I added the Ruby bin directory to the PATH. However the change does not take affect for the which_all command. Do you see anything wrong?

Edit:

Ah, I see which_all uses ORIGINAL_PATHS...

File: Library/Homebrew/requirements/ruby_requirement.rb [hunk 3]
```diff
@@ -9,23 +9,17 @@ def initialize(tags)
...
-          File.join(rubyhdrdir, arch),
-        ]
-        next unless includedirs.map {|dir| File.file?(File.join(dir, "ruby.h"))}.any?
-      elsif version == "1.8"
-        archdir = Utils.popen_read(ruby, "-rrbconfig", "-e", "print RbConfig::CONFIG['archdir']")
-        next unless File.file?(File.join(archdir, "ruby.h"))
-      end
-      Version.create(version) >= Version.create(@version)
+      rubyhdrdir = Utils.popen_read(ruby, "-rrbconfig", "-e", "print RbConfig::CONFIG['rubyhdrdir']")
+      next unless File.file?(File.join(rubyhdrdir, "ruby.h"))
```
Comment (sjackman) [inline Library/Homebrew/requirements/ruby_requirement.rb]: Try
```ruby
rubyhdrdir = Pathname.new Utils.popen_read(ruby, "-rrbconfig", "-e", "print RbConfig::CONFIG['rubyhdrdir']")
next unless (rubyhdrdir/"ruby.h").readable?
```

**Review by sjackman — CHANGES_REQUESTED** [review]

**Comment by carlorosati** [comment]
@sjackman 

prepend_path is ignoring new_enough_ruby since it is not a directory

See https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/ENV/shared.rb#L95

File: Library/Homebrew/requirements/ruby_requirement.rb [hunk 4]
```diff
@@ -14,11 +14,11 @@ def initialize(tags)
     path.unshift(RUBY_BIN) if which("ruby") != RUBY_PATH
     ENV["PATH"] = path.join(File::PATH_SEPARATOR)
 
-    which_all("ruby").detect do |ruby|
+    which_all("ruby").unshift(RUBY_PATH).detect do |ruby|
```
Comment (sjackman) [inline Library/Homebrew/requirements/ruby_requirement.rb]: This change seems redundant with line 14 above.
```ruby
path.unshift(RUBY_BIN) if which("ruby") != RUBY_PATH
```

**Review by sjackman — CHANGES_REQUESTED** [review]
Please squash your 6 commits to 1 and rebase onto `master`.

**Comment by sjackman** [comment]
> Ah, I see which_all uses `ORIGINAL_PATHS`...

`which_all` uses `PATH`, no?
https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/utils.rb#L314

**Comment by carlorosati** [comment]
Yes but requirement.rb overrides it on line 154

https://github.com/Linuxbrew/brew/blob/master/Library/Homebrew/requirement.rb#L154

**Comment by sjackman** [comment]
Ergh. I wonder why it does that. In any case, please squash and rebase.

**Comment by carlorosati** [comment]
@sjackman how's that?

File: Library/Homebrew/requirements/ruby_requirement.rb [hunk 5]
```diff
@@ -11,7 +11,7 @@ def initialize(tags)
   satisfy(build_env: false) { new_enough_ruby }
 
   env do
-    ENV.prepend_path "PATH", new_enough_ruby
+    ENV.prepend_path "PATH", new_enough_ruby.dirname
```
Comment (sjackman) [inline Library/Homebrew/requirements/ruby_requirement.rb:14]: What's the purpose of this change?

File: Library/Homebrew/requirements/ruby_requirement.rb [hunk 6]
```diff
@@ -44,6 +44,7 @@ def rubies
     if ruby_formula && ruby_formula.installed?
       rubies.unshift ruby_formula.bin/"ruby"
     end
+    rubies.unshift RUBY_PATH unless rubies.include? RUBY_PATH
```
Comment (sjackman) [inline Library/Homebrew/requirements/ruby_requirement.rb]: I prefer `rubies.push` over `rubies.unshift` to make `RUBY_PATH` a fallback plan rather than the first choice.

**Review by sjackman — CHANGES_REQUESTED** [review]
Where did the check for `ruby.h` go?

File: Library/Homebrew/requirements/ruby_requirement.rb (hunk 5 above)
Comment (carlorosati) [reply Library/Homebrew/requirements/ruby_requirement.rb:14]: prepend_path is ignoring new_enough_ruby since it is not a directory

See https://github.com/Homebrew/brew/blob/master/Library/Homebrew/extend/ENV/shared.rb#L95

Comment (carlorosati) [reply Library/Homebrew/requirements/ruby_requirement.rb:14]: I submitted a pull request upstream

Comment (sjackman) [reply Library/Homebrew/requirements/ruby_requirement.rb:14]: :+1: Thanks for the explanation.

File: Library/Homebrew/requirements/ruby_requirement.rb (hunk 6 above)
Comment (carlorosati) [reply Library/Homebrew/requirements/ruby_requirement.rb]: I like the way you think ;)

**Comment by carlorosati** [comment]
@sjackman I squashed the ruby.h checks. I put them back now.

**Comment by sjackman** [comment]
The CI failure is due to a bug that I've fixed just now. Can you please rebase onto `master`?

**Review by sjackman — APPROVED** [review]
Thanks for your contribution to Linuxbrew, Carlo! And thanks for sticking with this PR.
